server/rating_service/rating_service.py

```diff
@@ -148,192 +143,353 @@ async def _handle_rating_queue(self) -> None:
+            # Apply rating changes to global rating
+            new_global = Rating(
+                mu=global_rating.mu * mean_factor,
+                sigma=global_rating.sigma * dev_factor
+            )
+            # Only make the adjustment if the displayed rating increases
+            if displayed_rating(new_global) > displayed_rating(global_rating):
+                new_global_ratings[player_id] = new_global
+
+        return new_global_ratings
 
+    def _get_rating_data(
+        self,
+        conn,
+        summary: GameRatingSummary,
+        ratings: Dict[PlayerID, Rating]
+    ) -> GameRatingData:
         return [
             TeamRatingData(
                 team.outcome,
-                {player_id: ratings[player_id] for player_id in team.player_ids},
+                {
+                    player_id: ratings[player_id]
+                    for player_id in team.player_ids
+                },
             )
             for team in summary.teams
         ]
 
-    async def _get_player_rating(
-        self, player_id: int, rating_type: str, conn=None
-    ) -> Rating:
+    async def _get_players_rating_initialized(
+        self,
+        conn,
+        player_ratings: Dict[PlayerID, PlayerRatings],
+        rating_type: str
+    ) -> Dict[PlayerID, Rating]:
+        """
+        Initialize and return one rating type for all players
+        """
+        uninitialized_ratings = [
+            # Querying the key will create the value using rating
+            # initialization, sort of like a defaultdict.
+            (player_id, *player_ratings[player_id][rating_type])
+            for player_id in player_ratings.keys()
+            if rating_type not in player_ratings[player_id]
+        ]
+        if uninitialized_ratings:
+            # These players have not played any games using this leaderboard
+            # yet, so we need to create the initial ratings. This also
+            # ensures that the journal will show accurate changes.
+            await self._create_initial_ratings(
+                conn,
+                rating_type,
+                uninitialized_ratings
+            )
+
+        return {
+            player_id: Rating(*player_ratings[player_id][rating_type])
+            for player_id in player_ratings.keys()
+        }
+
+    async def _create_initial_ratings(
+        self,
+        conn,
+        rating_type: str,
+        ratings: Iterable[Tuple[PlayerID, float, float]]
+    ):
         if self._rating_type_ids is None:
             self._logger.warning(
-                "Tried to fetch player data before initializing service."
+                "Tried to write player data before initializing service."
             )
             raise ServiceNotReadyError("RatingService not yet initialized.")
 
-        rating_type_id = self._rating_type_ids.get(rating_type)
-        if rating_type_id is None:
-            raise ValueError(f"Unknown rating type {rating_type}.")
```

Comment:
I'm not sure it really matters. It just means we'll raise a `KeyError` on this line `self._rating_type_ids[rating_type]` instead of raising a `ValueError`. I guess it was there for nicer logging in the exception handler. I suppose I could put it back in.

Reply:
I just refactored the checking to happen at the very start of `_rate`. Since the other functions are private we can expect them to be called only in the places where we call them internally, so we know that this check will always have happened.